libs/execution_context/ebpf_program.c

```diff
@@ -1052,51 +1101,63 @@ _ebpf_program_get_helper_function_address(
     }
     provider_data_referenced = true;
 
-    if (helper_function_id > EBPF_MAX_GENERAL_HELPER_FUNCTION) {
-        if (!program->trampoline_table) {
-            return_value = EBPF_INVALID_ARGUMENT;
-            goto Done;
-        }
+    use_trampoline = program->parameters.code_type == EBPF_CODE_JIT;
+    ebpf_assert(program->trampoline_table != NULL || !use_trampoline);
+
+    // First check the trampoline table for the helper function.
+    if (use_trampoline) {
         return_value = ebpf_get_trampoline_function(program->trampoline_table, helper_function_id, &function_address);
-        if (return_value != EBPF_SUCCESS) {
-            goto Done;
+        if (return_value == EBPF_SUCCESS) {
+            found = true;
         }
+    }
 
-        *address = (uint64_t)function_address;
-    } else {
-        // First check if the extension has provided a new implementation of
-        // the global helper function.
-        if (program->trampoline_table) {
-            return_value =
-                ebpf_get_trampoline_function(program->trampoline_table, helper_function_id, &function_address);
-            if (return_value == EBPF_SUCCESS) {
-                // Found an entry.
-                *address = (uint64_t)function_address;
-                return_value = EBPF_SUCCESS;
-                goto Done;
+    // Check the general helper function table of the program type.
+    if (!found) {
+        for (size_t index = 0; index < program_data->program_info->count_of_global_helpers; index++) {
+            if (program_data->program_info->global_helper_prototype[index].helper_id == helper_function_id) {
```

Comment:
Instead of going over _both_ `global_helper_prototype` and `program_type_specific_helper_prototype` (at line 1129), we can check the `helper_id` against `EBPF_MAX_GENERAL_HELPER_FUNCTION` and decide to iterate only over one array.